homeassistant/components/netatmo/netatmo_entity_base.py

```diff
@@ -0,0 +1,127 @@
+    """Netatmo entity base class."""
+
+    DOMAIN = ""
+    TYPE = ""
+
+    def __init__(self, data_handler: NetatmoDataHandler) -> None:
+        """Set up Netatmo entity base."""
+        self.data_handler = data_handler
+        self._data_classes: List[Dict] = []
+
+        self._device_name = None
+        self._id = None
+        self._model = None
+        self._name = None
+        self._unique_id = None
+
+    async def async_added_to_hass(self) -> None:
+        """Entity created."""
+        _LOGGER.debug("New client %s", self.entity_id)
+        for data_class in self._data_classes:
+            if "home_id" in data_class:
+                await self.data_handler.register_data_class(
+                    data_class["name"], home_id=data_class["home_id"]
+                )
+                signal_name = f"{data_class['name']}-{data_class['home_id']}"
+            elif data_class["name"] == "PublicData":
+                await self.data_handler.register_data_class(
+                    data_class["name"],
+                    LAT_NE=data_class["LAT_NE"],
+                    LON_NE=data_class["LON_NE"],
+                    LAT_SW=data_class["LAT_SW"],
+                    LON_SW=data_class["LON_SW"],
+                    area_name=data_class["area_name"],
+                )
+                signal_name = f"{data_class['name']}-{data_class['area_name']}"
+            else:
+                await self.data_handler.register_data_class(data_class["name"])
+                signal_name = f"{data_class['name']}"
+
+            self.data_handler.listeners.append(
+                async_dispatcher_connect(
+                    self.hass,
+                    f"netatmo-update-{signal_name}",
+                    self.async_update_callback,
+                )
+            )
+        self.async_update_callback()
+
+    async def async_remove(self):
+        """Clean up when removing entity.
+
+        Remove entity if no entry in entity registry exist.
+        Remove entity registry entry if no entry in device registry exist.
+        Remove device registry entry if there is only one linked entity (this entity).
+        Remove entity registry entry if there are more than one entity linked to the
+        device registry entry.
+        """
+        entity_registry = await self.hass.helpers.entity_registry.async_get_registry()
+        entity_entry = entity_registry.async_get(self.entity_id)
+        if not entity_entry:
+            await super().async_remove()
+            return
+
+        device_registry = await self.hass.helpers.device_registry.async_get_registry()
+        device_entry = device_registry.async_get(entity_entry.device_id)
+        if not device_entry:
+            entity_registry.async_remove(self.entity_id)
+            return
+
+        if len(async_entries_for_device(entity_registry, entity_entry.device_id)) == 1:
+            device_registry.async_remove_device(device_entry.id)
+            return
```

Comment:
You might be able to remove this

I'm pretty sure we clean up the device register automatically now